### PR TITLE
#757 Add Unique Transaction ID to the logs

### PR DESCRIPTION
**Pull request by Bubomira**
# Add Unique Transaction ID to the logs

create a unique id for each transaction
logs with more meaningful messages + transaction id for the existing logs

### CODE REVIEW

File: services/py-api/poetry.lock
Comment (IvanObreshkov) [inline services/py-api/poetry.lock:1]: remove this

File: services/py-api/pyproject.toml
Comment (IvanObreshkov) [inline services/py-api/pyproject.toml:1]: remove this

File: services/py-api/src/database/repository/participants_repository.py [hunk 1]
```diff
@@ -48,12 +46,29 @@ async def create(
...
                 team_id=kwargs.get("team_id"),
             )
-            LOG.debug("Inserting participant...", participant=participant.dump_as_json())
+            if session:
+                LOG.debug(
+                    "Inserting participant {participant} via transaction {id}",
+                    participant=participant.dump_as_json(),
+                    id=session.transaction_id,
+                )
+
```
Comment (IvanObreshkov) [inline services/py-api/src/database/repository/participants_repository.py:55]: Where is the normal log, when session is None?

File: services/py-api/src/database/repository/participants_repository.py [hunk 2]
```diff
@@ -48,12 +46,29 @@ async def create(
...
             return Ok(participant)
         except DuplicateKeyError:
-            LOG.warning("Participant insertion failed due to duplicate email", email=input_data.email)
+            if session:
+                LOG.warning(
+                    "Participant insertion failed due to duplicate email {email} via transaction {id}",
+                    email=input_data.email,
+                    transaction_id=session.transaction_id,
+                )
+
```
Comment (IvanObreshkov) [inline services/py-api/src/database/repository/participants_repository.py:65]: Where is the normal log, when session is None?

File: services/py-api/src/database/repository/participants_repository.py [hunk 3]
```diff
@@ -48,12 +46,29 @@ async def create(
...
+                    id=session.transaction_id,
+                )
+
             await self._collection.insert_one(document=participant.dump_as_mongo_db_document(), session=session)
             return Ok(participant)
         except DuplicateKeyError:
-            LOG.warning("Participant insertion failed due to duplicate email", email=input_data.email)
+            if session:
+                LOG.warning(
+                    "Participant insertion failed due to duplicate email {email} via transaction {id}",
```
Comment (IvanObreshkov) [inline services/py-api/src/database/repository/participants_repository.py:61]: You do not need the things in the {}, they are not replaced with the values, they just sit there as strings

File: services/py-api/src/database/repository/participants_repository.py [hunk 4]
```diff
@@ -48,12 +46,29 @@ async def create(
                 # If the team_id is passed as a kwarg the participant will be inserted in the given team
                 team_id=kwargs.get("team_id"),
             )
-            LOG.debug("Inserting participant...", participant=participant.dump_as_json())
+            if session:
+                LOG.debug(
+                    "Inserting participant {participant} via transaction {id}",
```
Comment (IvanObreshkov) [inline services/py-api/src/database/repository/participants_repository.py:51]: You do not need the things in the {}, they are not replaced with the values, they just sit there as strings

File: services/py-api/src/database/repository/participants_repository.py [hunk 5]
```diff
@@ -48,12 +46,29 @@ async def create(
...
+
             return Err(DuplicateEmailError(input_data.email))
         except Exception as e:
-            LOG.exception("Participant insertion failed due to err {}".format(e))
+            if session:
+                LOG.exception(
+                    "Participant insertion failed due to err {e} via transaction {id}",
+                    e=e,
+                    transaction_id=session.transaction_id,
+                )
```
Comment (IvanObreshkov) [inline services/py-api/src/database/repository/participants_repository.py:73]: Where is the normal log, when session is None?

File: services/py-api/src/database/repository/participants_repository.py [hunk 6]
```diff
@@ -48,12 +46,29 @@ async def create(
...
+                    email=input_data.email,
+                    transaction_id=session.transaction_id,
+                )
+
             return Err(DuplicateEmailError(input_data.email))
         except Exception as e:
-            LOG.exception("Participant insertion failed due to err {}".format(e))
+            if session:
+                LOG.exception(
+                    "Participant insertion failed due to err {e} via transaction {id}",
```
Comment (IvanObreshkov) [inline services/py-api/src/database/repository/participants_repository.py:70]: You do not need the things in the {}, they are not replaced with the values, they just sit there as strings

File: services/py-api/src/database/repository/teams_repository.py [hunk 7]
```diff
@@ -20,24 +21,32 @@ def __init__(self, db_manager: DatabaseManager, collection_name: str) -> None:
...
+        self, input_data: ParticipantRequestBody, session: Optional[CustomSession] = None, **kwargs: Dict[str, Any]
     ) -> Result[Team, DuplicateTeamNameError | Exception]:
 
         try:
             team = Team(name=input_data.team_name)
-            LOG.debug("Inserting team...", team=team.dump_as_json())
+            if session:
+                LOG.debug(
+                    "Inserting team {team} via transaction {id}", team=team.dump_as_json(), id=session.transaction_id
+                )
```
Comment (IvanObreshkov) [inline services/py-api/src/database/repository/teams_repository.py:32]: again

File: services/py-api/src/database/repository/teams_repository.py [hunk 8]
```diff
@@ -20,24 +21,32 @@ def __init__(self, db_manager: DatabaseManager, collection_name: str) -> None:
...
             return Ok(team)
 
         except DuplicateKeyError:
-            LOG.warning("Team insertion failed due to duplicate team name", team_name=input_data.team_name)
+            if session:
+                LOG.warning(
+                    "Team insertion failed due to duplicate team name {team_name} via transaction {id}",
+                    team_name=input_data.team_name,
+                    id=session.transaction_id,
+                )
```
Comment (IvanObreshkov) [inline services/py-api/src/database/repository/teams_repository.py:42]: and again

File: services/py-api/src/database/repository/teams_repository.py [hunk 9]
```diff
@@ -20,24 +21,32 @@ def __init__(self, db_manager: DatabaseManager, collection_name: str) -> None:
...
             return Err(DuplicateTeamNameError(input_data.team_name))
 
         except Exception as e:
-            LOG.exception("Team insertion failed due to err {}".format(e))
+            if session:
+                LOG.exception(
+                    "Team insertion failed due to err {e} via transaction {id}", e=e, id=session.transaction_id
+                )
```
Comment (IvanObreshkov) [inline services/py-api/src/database/repository/teams_repository.py:49]: and again

**Comment by asynchroza** [comment]
@IvanObreshkov It seems that the main issue here is that Lyubomira is unsure how to manually validate the changes she has pushed. Ideally, we shouldn’t write code and simply hope it works. Please involve @Isak-Bego and discuss a proper approach to testing your code, even if some parts of the implementation are incomplete.

If the code cannot currently be tested, consider delaying its development until it can be, or focusing on tasks with higher priority and easier testability. Remember, all code can be tested, but if a lack of knowledge is the barrier to testing, it might be more efficient to redirect efforts towards tasks that allow for easier validation and higher overall importance. By doing so, developers can gradually learn how to validate their changes even when a clear testing path is not provided. What I'm trying to say is... avoid putting them in situations where they require constant involvement from you.

@Bubomira  As members of the department, it’s our responsibility to thoroughly test our own changes. Team leads are going above and beyond to review this code. Context switching is one of the most challenging aspects of being a team lead. Keeping track of everyone’s issues, considering all edge cases, and meticulously reviewing for both logical and non-logical errors makes it nearly impossible to operate at 100% efficiency every time they review someone’s changes.

In the long run, this lack of testing often ends up causing more issues for you than anyone else. Use this as an opportunity to learn and grow. By understanding the code and prioritizing testing, you’re not only helping yourself but also supporting the team. Let’s make this a priority moving forward.

Nevertheless, I'm really happy to see you putting in effort and being involved with @IvanObreshkov's ugly python. I'm sure if you involve them in a chat, they will be more than willing to help you figure out how to test your changes.

But mostly, @Isak-Bego @IvanObreshkov I suggest that you ditch this story for now. You should first teach them how to inspect the logs, what's their purpose, show them log files and how they've helped you debug issues. I'm pretty sure none of this really makes sense in their heads right now. It's easy for you now but it wasn't back then. Be the mentor you needed.

**Comment by Isak-Bego** [comment]
> @IvanObreshkov It seems that the main issue here is that Lyubomira is unsure how to manually validate the changes she has pushed. Ideally, we shouldn’t write code and simply hope it works. Please involve @Isak-Bego and discuss a proper approach to testing your code, even if some parts of the implementation are incomplete.
> 
> If the code cannot currently be tested, consider delaying its development until it can be, or focusing on tasks with higher priority and easier testability. Remember, all code can be tested, but if a lack of knowledge is the barrier to testing, it might be more efficient to redirect efforts towards tasks that allow for easier validation and higher overall importance. By doing so, developers can gradually learn how to validate their changes even when a clear testing path is not provided. What I'm trying to say is... avoid putting them in situations where they require constant involvement from you.
> 
> @Bubomira As members of the department, it’s our responsibility to thoroughly test our own changes. Team leads are going above and beyond to review this code. Context switching is one of the most challenging aspects of being a team lead. Keeping track of everyone’s issues, considering all edge cases, and meticulously reviewing for both logical and non-logical errors makes it nearly impossible to operate at 100% efficiency every time they review someone’s changes.
> 
> In the long run, this lack of testing often ends up causing more issues for you than anyone else. Use this as an opportunity to learn and grow. By understanding the code and prioritizing testing, you’re not only helping yourself but also supporting the team. Let’s make this a priority moving forward.
> 
> Nevertheless, I'm really happy to see you putting in effort and being involved with @IvanObreshkov's ugly python. I'm sure if you involve them in a chat, they will be more than willing to help you figure out how to test your changes.
> 
> But mostly, @Isak-Bego @IvanObreshkov I suggest that you ditch this story for now. You should first teach them how to inspect the logs, what's their purpose, show them log files and how they've helped you debug issues. I'm pretty sure none of this really makes sense in their heads right now. It's easy for you now but it wasn't back then. Be the mentor you needed.

@asynchroza Thank you for making such good points. We will try to improve and give clearer instructions for the new contributors. We have been trying to be available for them at all times. I think I discussed it briefly with @IvanObreshkov and it is better to postpone this for now as it does not have such a high priority. Your feedback gives us a clearer view of the situation and helps us switch to a more beginner-friendly methodology.

**Comment by iobreshkov-qb** [comment]
> @IvanObreshkov It seems that the main issue here is that Lyubomira is unsure how to manually validate the changes she has pushed. Ideally, we shouldn’t write code and simply hope it works. Please involve @Isak-Bego and discuss a proper approach to testing your code, even if some parts of the implementation are incomplete.
> 
> If the code cannot currently be tested, consider delaying its development until it can be, or focusing on tasks with higher priority and easier testability. Remember, all code can be tested, but if a lack of knowledge is the barrier to testing, it might be more efficient to redirect efforts towards tasks that allow for easier validation and higher overall importance. By doing so, developers can gradually learn how to validate their changes even when a clear testing path is not provided. What I'm trying to say is... avoid putting them in situations where they require constant involvement from you.
> 
> @Bubomira As members of the department, it’s our responsibility to thoroughly test our own changes. Team leads are going above and beyond to review this code. Context switching is one of the most challenging aspects of being a team lead. Keeping track of everyone’s issues, considering all edge cases, and meticulously reviewing for both logical and non-logical errors makes it nearly impossible to operate at 100% efficiency every time they review someone’s changes.
> 
> In the long run, this lack of testing often ends up causing more issues for you than anyone else. Use this as an opportunity to learn and grow. By understanding the code and prioritizing testing, you’re not only helping yourself but also supporting the team. Let’s make this a priority moving forward.
> 
> Nevertheless, I'm really happy to see you putting in effort and being involved with @IvanObreshkov's ugly python. I'm sure if you involve them in a chat, they will be more than willing to help you figure out how to test your changes.
> 
> But mostly, @Isak-Bego @IvanObreshkov I suggest that you ditch this story for now. You should first teach them how to inspect the logs, what's their purpose, show them log files and how they've helped you debug issues. I'm pretty sure none of this really makes sense in their heads right now. It's easy for you now but it wasn't back then. Be the mentor you needed.

Thank you for the feedback, really appreciate it. As @Isak-Bego mentioned, this is not really a high-priority ticket, but I think we have discussed with @Bubomira how this should be implemented and what is the purpose of it. We could ditch the ticket for later.

**Pull request closed without merging**